In [ ]:
import os
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [ ]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [ ]:
class CrimeHotspot():
  def get_null_count(self, df):
    column_names = list(df.columns)
    null_values = list(df.isnull().sum().values)
    ser = pd.Series(null_values,column_names)
    null_dict = {}
    for key, value in ser.items():
      if value !=0:
        null_dict[key] = value
    return null_dict

  def rename_col(self, data):
    column_mapping = {
    'index': 'Index',
    'ARREST_KEY': 'Arrest_ID',
    'ARREST_DATE': 'Arrest_Date',
    'PD_CD': 'PD_Code',
    'PD_DESC': 'PD_Description',
    'KY_CD': 'KY_Code',
    'OFNS_DESC': 'Offense_Description',
    'LAW_CODE': 'Law_Code',
    'LAW_CAT_CD': 'Law_Category',
    'ARREST_BORO': 'Arrest_Borough',
    'ARREST_PRECINCT': 'Arrest_Precinct',
    'JURISDICTION_CODE': 'Jurisdiction_Code',
    'AGE_GROUP': 'Age_Group',
    'PERP_SEX': 'Perpetrator_Sex',
    'PERP_RACE': 'Perpetrator_Race',
    'X_COORD_CD': 'X_Coordinate',
    'Y_COORD_CD': 'Y_Coordinate',
    'Latitude': 'Latitude',
    'Longitude': 'Longitude',
    'New Georeferenced Column': 'Georeferenced_Location'
      }
    data.rename(columns=column_mapping, inplace=True)
    return data


  def delete_col(self, data):
    columns_to_delete = ['PD_Code', 'PD_Description', 'KY_Code', 'Jurisdiction_Code']
    for col in columns_to_delete:
        if col in data.columns:
            del data[col]
    return data


  def get_info(self, dataframe):
    dataframe = dataframe.reset_index()
    print("Basic information about dataset")
    print("-------------------------------------------------------------------------------------------")
    print(f"Number of Rows : {len(dataframe.index)}", end='\n\n')
    print(f"{dataframe.index}", end ="\n\n")
    print(f"Number of columns :{len(dataframe.columns)}", end='\n\n')
    print(f"column Name : {np.array(dataframe.columns)}", end='\n\n')
    print("Columns which has null values : ", self.get_null_count(dataframe), end='\n\n')
    print("-------------------------------------------------------------------------------------------")



  def create_col(self, data):
    offense_data = {
        '260.00': ['Abandonment of a child', 'E Felony'],
        '125.45': ['Abortion in the first degree', 'D Felony'],
        '125.40': ['Abortion in the second degree', 'E Felony'],
        '205.19': ['Absconding from a community treatment facility', 'E Felony'],
        '205.18': ['Absconding from a furlough program', 'A Misdemeanor'],
        '205.17': ['Absconding from temporary release in the first degree', 'E Felony'],
        '255.17': ['Adultery', 'B Misdemeanor'],
        '275.30': ['Advertisement or sale of unauthorized recordings in the first degree', 'E Felony'],
        '275.25': ['Advertisement or sale of unauthorized recordings in the second degree', 'A Misdemeanor'],
        '120.12': ['Aggravated assault upon a person less than eleven years old', 'E Felony'],
        '120.11': ['Aggravated assault upon a police officer or a peace officer', 'B Violent Felony'],
        '145.27': ['Aggravated cemetery desecration in the first degree', 'D Felony'],
        '145.26': ['Aggravated cemetery desecration in the second degree', 'E Felony'],
        '215.52': ['Aggravated criminal contempt', 'D Felony'],
        '265.19': ['Aggravated criminal possession of a weapon', 'C Felony'],
        '222.65': ['Aggravated criminal sale of cannabis', 'C Felony'],
        '125.11': ['Aggravated criminally negligent homicide', 'C Violent Felony'],
        '460.22': ['Aggravated enterprise corruption', 'A-1 Felony'],
        '240.75': ['Aggravated family offense', 'E Felony'],
        '155.43': ['Aggravated grand larceny of an automated teller machine', 'C Felony'],
        '240.31': ['Aggravated harassment in the first degree', 'E Felony'],
        '240.30': ['Aggravated harassment in the second degree', 'A Misdemeanor'],
        '240.32': ['Aggravated harassment of an employee by an inmate', 'E Felony'],
        '190.80': ['Aggravated identity theft', 'D Felony'],
        '176.35': ['Aggravated insurance fraud', 'D Felony'],
        '240.73': ['Aggravated interference with health care services in the first degree', 'E Felony'],
        '240.72': ['Aggravated interference with health care services in the second degree', 'E Felony'],
        '125.22': ['Aggravated manslaughter in the first degree', 'B Violent Felony'],
        '125.21': ['Aggravated manslaughter in the second degree', 'C Violent Felony'],
        '125.26': ['Aggravated murder', 'A-I Felony'],
        '130.70': ['Aggravated sexual abuse in the first degree', 'B Violent Felony'],
        '130.65': ['Aggravated sexual abuse in the fourth degree', 'E Violent Felony'],
        '130.67': ['Aggravated sexual abuse in the second degree', 'C Violent Felony'],
        '130.66': ['Aggravated sexual abuse in the third degree', 'D Violent Felony'],
        '405.18': ['Aggravated unpermitted use of indoor pyrotechnics in the first degree', 'D Violent Felony'],
        '405.16': ['Aggravated unpermitted use of indoor pyrotechnics in the second degree', 'E Felony'],
        '120.04': ['Aggravated vehicular assault', 'C Felony'],
        '125.14': ['Aggravated vehicular homicide', 'B Felony'],
        '240.40': ['Appearance in public under the influence of narcotics or a drug other than alcohol', 'Violation'],
        '150.01': ['Arson in the fifth degree', 'A Misdemeanor'],
        '150.20': ['Arson in the first degree', 'A-I Felony'],
        '150.05': ['Arson in the fourth degree', 'E Felony'],
        '150.15': ['Arson in the second degree', 'B Violent Felony'],
        '150.10': ['Arson in the third degree', 'C Felony'],
        '120.10': ['Assault in the first degree', 'B Violent Felony'],
        '120.05': ['Assault in the second degree', 'D Violent Felony'],
        '120.00': ['Assault in the third degree', 'A Misdemeanor'],
        '120.09': ['Assault on a judge', 'C Violent Felony'],
        '120.08': ['Assault on a peace officer police officer fireman or emergency medical services professional', 'C Violent Felony'],
        '165.11': ['Auto stripping in the first degree', 'D Felony'],
        '165.10': ['Auto stripping in the second degree', 'E Felony'],
        '165.09': ['Auto stripping in the third degree', 'A Misdemeanor'],
        "215.57": ["Bail jumping in the first degree", "D Felony"],
        "215.56": ["Bail jumping in the second degree", "E Felony"],
        "215.55": ["Bail jumping in the third degree", "A Misdemeanor"],
        "255.15": ["Bigamy", "E Felony"],
        "200.45": ["Bribe giving for public office", "D Felony"],
        "215.20": ["Bribe receiving by a juror", "D Felony"],
        "180.25": ["Bribe receiving by a labor official", "D Felony"],
        "215.05": ["Bribe receiving by a witness", "D Felony"],
        "200.50": ["Bribe receiving for public office", "D Felony"],
        "200.12": ["Bribe receiving in the first degree", "B Felony"],
        "200.11": ["Bribe receiving in the second degree", "C Felony"],
        "200.10": ["Bribe receiving in the third degree", "D Felony"],
        "200.04": ["Bribery in the first degree", "B Felony"],
        "200.03": ["Bribery in the second degree", "C Felony"],
        "200.00": ["Bribery in the third degree", "D Felony"],
        "215.19": ["Bribing a juror", "D Felony"],
        "180.15": ["Bribing a labor official", "D Felony"],
        "140.30": ["Burglary in the first degree", "B Violent Felony"],
        "140.25": ["Burglary in the second degree", "C Violent Felony"],
        "140.20": ["Burglary in the third degree", "D Felony"],
        "265.26": ["Burn injury and wounds to be reported", "A Misdemeanor"],
        "145.23": ["Cemetery desecration in the first degree", "E Felony"],
        "145.22": ["Cemetery desecration in the second degree", "A Misdemeanor"],
        "265.25": ["Certain wounds to be reported", "A Misdemeanor"],
        "135.36": ["Coercion in the first degree", "D Felony"],
        "135.60": ["Coercion in the second degree", "A Misdemeanor"],
        "180.08": ["Commercial bribe receiving in the first degree", "E Felony"],
        "180.05": ["Commercial bribe receiving in the second degree", "A Misdemeanor"],
        "180.03": ["Commercial bribing in the first degree", "E Felony"],
        "180.00": ["Commercial bribing in the second degree", "A Misdemeanor"],
        "230.33": ["Compelling prostitution", "B Felony"],
        "215.45": ["Compounding a crime", "A Misdemeanor"],
        "156.27": ["Computer tampering in the first degree", "C Felony"],
        "156.20": ["Computer tampering in the fourth degree", "A Misdemeanor"],
        "156.26": ["Computer tampering in the second degree", "D Felony"],
        "156.25": ["Computer tampering in the third degree", "E Felony"],
        "156.10": ["Computer trespass", "E Felony"],
        "105.05": ["Conspiracy in the fifth degree", "A Misdemeanor"],
        "105.17": ["Conspiracy in the first degree", "A-I Felony"],
        "105.10": ["Conspiracy in the fourth degree", "E Felony"],
        "105.15": ["Conspiracy in the second degree", "B Felony"],
        "105.00": ["Conspiracy in the sixth degree", "B Misdemeanor"],
        "105.13": ["Conspiracy in the third degree", "D Felony"],
        "496.05": ["Corrupting the government in the first degree", "B Felony"],
        "496.02": ["Corrupting the government in the fourth degree", "E Felony"],
        "496.04": ["Corrupting the government in the second degree", "C Felony"],
        "496.03": ["Corrupting the government in the third degree", "D Felony"],
        "130.75": ["Course of sexual conduct against a child in the first degree", "B Violent Felony"],
        "130.80": ["Course of sexual conduct against a child in the second degree", "D Violent Felony"],
        "270.10": ["Creating a hazard", "B Misdemeanor"],
        "490.25": ["Crime of terrorism", "A-I Felony"],
        "220.70": ["Criminal possession of methamphetamine manufacturing material in the second degree", "A Misdemeanor"],
        "240.15": ["Criminal anarchy", "E Felony"],
        "215.51": ["Criminal contempt in the first degree", "E Felony"],
        "215.50": ["Criminal contempt in the second degree", "A Misdemeanor"],
        "215.65": ["Criminal contempt of a temporary state commission", "A Misdemeanor"],
        "215.60": ["Criminal contempt of the legislature", "A Misdemeanor"],
        "215.66": ["Criminal contempt of the state commission on judicial conduct", "A Misdemeanor"],
        "179.10": ["Criminal diversion of medical marihuana in the first degree", "E Felony"],
        "179.11": ["Criminal diversion of medical marihuana in the second degree", "B Misdemeanor"],
        "178.25": ["Criminal diversion of prescription medications and prescriptions in the first degree", "C Felony"],
        "178.10": ["Criminal diversion of prescription medications and prescriptions in the fourth degree", "A Misdemeanor"],
        "178.15": ['Criminal diversion of prescription medications and prescriptions in the third degree', 'E Felony'],
        "115.08": ['Criminal facilitation in the first degree', 'B Felony'],
        "115.00": ['Criminal facilitation in the fourth degree', 'A Misdemeanor'],
        "115.05": ['Criminal facilitation in the second degree', 'C Felony'],
        "115.01": ['Criminal facilitation in the third degree', 'E Felony'],
        "190.26": ['Criminal impersonation in the first degree', 'E Felony'],
        "190.25": ['Criminal impersonation in the second degree', 'A Misdemeanor'],
        "220.46": ['Criminal injection of a narcotic drug', 'E Felony'],
        "240.71": ['Criminal interference with health care services or religious worship in the first degree', 'E Felony'],
        "240.70": ['Criminal interference with health care services or religious worship in the second degree', 'A Misdemeanor'],
        "145.12": ['Criminal mischief in the first degree', 'B Felony'],
        "145.00": ['Criminal mischief in the fourth degree', 'A Misdemeanor'],
        "145.10": ['Criminal mischief in the second degree', 'D Felony'],
        "145.05": ['Criminal mischief in the third degree', 'E Felony'],
        "240.46": ['Criminal nuisance in the first degree', 'E Felony'],
        "240.45": ['Criminal nuisance in the second degree', 'B Misdemeanor'],
        "121.11": ['Criminal obstruction of breathing or blood circulation', 'A Misdemeanor'],
        "220.71": ['Criminal possession of methamphetamine manufacturing material in the first degree', 'E Felony'],
        "490.45": ['Criminal possession of a chemical weapon or biological weapon in the first degree', 'A-I Felony'],
        "490.40": ['Criminal possession of a chemical weapon or biological weapon in the second degree', 'B Violent Felony'],
        "490.37": ['Criminal possession of a chemical weapon or biological weapon in the third degree', 'C Violent Felony'],
        "220.06": ['Criminal possession of a controlled substance in the fifth degree', 'D Felony'],
        "220.21": ['Criminal possession of a controlled substance in the first degree', 'A-I Felony'],
        "220.09": ['Criminal possession of a controlled substance in the fourth degree', 'C Felony'],
        "220.18": ['Criminal possession of a controlled substance in the second degree', 'A-II Felony'],
        "220.03": ['Criminal possession of a controlled substance in the seventh degree', 'A Misdemeanor'],
        "220.16": ['Criminal possession of a controlled substance in the third degree', 'B Felony'],
        "265.1B": ['Criminal possession of a firearm', 'E Felony'],
        "170.30": ['Criminal possession of a forged instrument in the first degree', 'C Felony'],
        "170.25": ['Criminal possession of a forged instrument in the second degree', 'D Felony'],
        "170.20": ['Criminal possession of a forged instrument in the third degree', 'A Misdemeanor'],
        "145.70": ['Criminal possession of a taximeter accelerating device', 'A Misdemeanor'],
        "265.04": ['Criminal possession of a weapon in the first degree', 'B Violent Felony'],
        "265.01": ['Criminal possession of a weapon in the fourth degree', 'A Misdemeanor'],
        "265.03": ['Criminal possession of a weapon in the second degree', 'C Violent Felony'],
        "265.02": ['Criminal possession of a weapon in the third degree', 'D Violent Felony'],
        "265.1A": ['Criminal possession of a weapon on school grounds', 'E Felony'],
        "170.47": ['Criminal possession of an anti-security item', 'B Misdemeanor'],
        "222.40": ['Criminal possession of cannabis in the first degree', 'D Felony'],
        "222.35": ['Criminal possession of cannabis in the second degree', 'E Felony'],
        "222.30": ['Criminal possession of cannabis in the third degree', 'A Misdemeanor'],
        "156.35": ['Criminal possession of computer related material', 'E Felony'],
        "170.40": ['Criminal possession of forgery devices', 'D Felony'],
        "221.10": ['Criminal possession of marihuana in the fifth degree', 'B Misdemeanor'],
        "221.30": ['Criminal possession of marihuana in the first degree', 'C Felony'],
        "221.15": ['Criminal possession of marihuana in the fourth degree', 'A Misdemeanor'],
        "221.25": ['Criminal possession of marihuana in the second degree', 'D Felony'],
        "221.20": ['Criminal possession of marihuana in the third degree', 'E Felony'],
        "220.60": ['Criminal possession of precursors of controlled substances', 'E Felony'],
        "220.72": ['Criminal possession of precursors of methamphetamine', 'E Felony'],
        "158.50": ['Criminal possession of public benefit cards in the first degree', 'C Felony'],
        "158.45": ['Criminal possession of public benefit cards in the second degree', 'D Felony'],
        "158.40": ['Criminal possession of public benefit cards in the third degree', 'E Felony'],
        "165.40": ['Criminal possession of stolen property in the fifth degree', 'A Misdemeanor'],
        "165.54": ['Criminal possession of stolen property in the first degree', 'B Felony'],
        "165.45": ['Criminal possession of stolen property in the fourth degree', 'E Felony'],
        "165.52": ['Criminal possession of stolen property in the second degree', 'C Felony'],
        "165.50": ['Criminal possession of stolen property in the third degree', 'D Felony'],
        "265.17": ['Criminal purchase or disposal of a weapon', 'D Felony'],
        "179.15": ['Criminal retention of medical marihuana', 'A Misdemeanor'],
        "220.44": ['Criminal sale of a controlled substance in or near school grounds', 'B Felony'],
        "220.31": ['Criminal sale of a controlled substance in the fifth degree', 'D Felony'],
        "220.43": ['Criminal sale of a controlled substance in the first degree', 'A-I Felony'],
        "220.34": ['Criminal sale of a controlled substance in the fourth degree', 'C Felony'],
        "220.41": ['Criminal sale of a controlled substance in the second degree', 'A-II Felony'],
        "220.39": ['Criminal sale of a controlled substance in the third degree', 'B Felony'],
        "220.48": ['Criminal sale of a controlled substance to a child', 'B Felony'],
        "265.13": ['Criminal sale of a firearm in the first degree', 'B Violent Felony'],
        "265.12": ['Criminal sale of a firearm in the second degree', 'C Violent Felony'],
        "265.11": ['Criminal sale of a firearm in the third degree', 'D Violent Felony'],
        "265.16": ['Criminal sale of a firearm to a minor', 'C Felony'],
        "265.14": ['Criminal sale of a firearm with the aid of a minor', 'C Violent Felony'],
        "190.27": ['Criminal sale of a police uniform', 'A Misdemeanor'],
        "220.65": ['Criminal sale of a prescription for a controlled substance or of a controlled substance by a practitioner or pharmacist', 'C Felony'],
        "222.60": ['Criminal sale of cannabis in the first degree', 'D Felony'],
        "222.55": ['Criminal sale of cannabis in the second degree', 'E Felony'],
        "222.50": ['Criminal sale of cannabis in the third degree', 'A Misdemeanor'],
        "221.35": ['Criminal sale of marihuana in the fifth degree', 'B Misdemeanor'],
        "221.55": ['Criminal sale of marihuana in the first degree', 'C Felony'],
        "221.40": ['Criminal sale of marihuana in the fourth degree', 'A Misdemeanor'],
        "221.50": ['Criminal sale of marihuana in the second degree', 'D Felony'],
        "221.45": ['Criminal sale of marihuana in the third degree', 'E Felony'],
        "130.50": ['Criminal sexual act in the first degree', 'B Violent Felony'],
        "130.45": ['Criminal sexual act in the second degree', 'D Violent Felony'],
        "130.40": ['Criminal sexual act in the third degree', 'E Felony'],
        "170.45": ['Criminal simulation', 'A Misdemeanor'],
        "100.00": ['Criminal solicitation in the fifth degree', 'Violation'],
        "100.13": ['Criminal solicitation in the first degree', 'C Felony'],
        "100.05": ['Criminal solicitation in the fourth degree', 'A Misdemeanor'],
        "100.10": ['Criminal solicitation in the second degree', 'D Felony'],
        "100.08": ['Criminal solicitation in the third degree', 'E Felony'],
        "145.20": ['Criminal tampering in the first degree', 'D Felony'],
        "145.15": ['Criminal tampering in the second degree', 'A Misdemeanor'],
        "145.14": ['Criminal tampering in the third degree', 'B Misdemeanor'],
        "140.17": ['Criminal trespass in the first degree', 'D Felony'],
        "140.15": ['Criminal trespass in the second degree', 'A Misdemeanor'],
        "140.10": ['Criminal trespass in the third degree', 'B Misdemeanor'],
        "490.55": ['Criminal use of a chemical weapon or biological weapon in the first degree', 'A-I Felony'],
        "490.50": ['Criminal use of a chemical weapon or biological weapon in the second degree', 'A-II Felony'],
        "490.47": ['Criminal use of a chemical weapon or biological weapon in the third degree', 'B Violent Felony'],
        "265.09": ['Criminal use of a firearm in the first degree', 'B Violent Felony'],
        "265.08": ['Criminal use of a firearm in the second degree', 'C Violent Felony'],
        "158.35": ['Criminal use of a public benefit card in the first degree', 'E Felony'],
        "158.30": ['Criminal use of a public benefit card in the second degree', 'A Misdemeanor'],
        "190.76": ['Criminal use of an access device in the first degree', 'E Felony'],
        "190.75": ['Criminal use of an access device in the second degree', 'A Misdemeanor'],
        "190.42": ['Criminal usury in the first degree', 'C Felony'],
        "125.10": ['Criminally negligent homicide', 'E Felony'],
        "220.45": ['Criminally possessing a hypodermic instrument', 'A Misdemeanor'],
        "220.55": ['Criminally using drug paraphernalia in the first degree', 'D Felony'],
        "220.50": ['Criminally using drug paraphernalia in the second degree', 'A Misdemeanor'],
        "135.50": ['Custodial interference in the first degree', 'E Felony'],
        "135.45": ['Custodial interference in the second degree', 'A Misdemeanor'],
        "195.20": ['Defrauding the government', 'E Felony'],
        "240.77": ['Directing a laser at an aircraft in the first degree', 'E Felony'],
        "240.76": ['Directing a laser at an aircraft in the second degree', 'A Misdemeanor'],
        "240.20": ['Disorderly conduct', 'Violation'],
        "240.21": ['Disruption or disturbance of religious service', 'A Misdemeanor'],
        "240.48": ['Disseminating a false registered sex offender notice', 'A Misdemeanor'],
        "235.22": ['Disseminating indecent material to minors in the first degree', 'D Felony'],
        "235.21": ['Disseminating indecent material to minors in the second degree', 'E Felony'],
        "250.60": ['Dissemination of an unlawful surveillance image in the first degree', 'E Felony'],
        "250.55": ['Dissemination of an unlawful surveillance image in the second degree', 'A Misdemeanor'],
        "250.20": ['Divulging an eavesdropping warrant', 'A Misdemeanor'],
        "250.05": ["Eavesdropping", "E Felony"],
        "215.14": ["Employer unlawfully penalizing witness or victim", "B Misdemeanor"],
        "260.10": ["Endangering the welfare of a child", "A Misdemeanor"],
        "260.34": ["Endangering the welfare of a vulnerable elderly person, or an incompetent or physically disabled person in the first degree", "D Felony"],
        "260.32": ["Endangering the welfare of a vulnerable elderly person, or an incompetent or physically disabled person in the second degree", "E Felony"],
        "260.25": ["Endangering the welfare of an incompetent or physically disabled person in the first degree", "E Felony"],
        "260.24": ["Endangering the welfare of an incompetent or physically disabled person in the second degree", "A Misdemeanor"],
        "460.20": ["Enterprise corruption", "B Felony"],
        "205.15": ["Escape in the first degree", "D Felony"],
        "205.10": ["Escape in the second degree", "E Felony"],
        "205.05": ["Escape in the third degree", "A Misdemeanor"],
        "245.01": ["Exposure of a person", "Violation"],
        "130.90": ["Facilitating a sex offense with a controlled substance", "D Violent Felony"],
        "263.30": ["Facilitating a sexual performance by a child with a controlled substance or alcohol", "B Felony"],
        "250.35": ["Failing to report criminal communications", "B Misdemeanor"],
        "215.58": ["Failing to respond to an appearance ticket", "Violation"],
        "275.40": ["Failure to disclose the origin of a recording in the first degree", "E Felony"],
        "275.35": ["Failure to disclose the origin of a recording in the second degree", "A Misdemeanor"],
        "250.15": ["Failure to report wiretapping", "B Misdemeanor"],
        "190.20": ["False advertising", "A Misdemeanor"],
        "190.23": ["False personation", "B Misdemeanor"],
        "240.60": ["Falsely reporting an incident in the first degree", "D Violent Felony"],
        "240.55": ["Falsely reporting an incident in the second degree", "E Violent Felony"],
        "240.50": ["Falsely reporting an incident in the third degree", "A Misdemeanor"],
        "175.10": ["Falsifying business records in the first degree", "E Felony"],
        "175.05": ["Falsifying business records in the second degree", "A Misdemeanor"],
        "130.85": ["Female genital mutilation", "E Felony"],
        "130.52": ["Forcible touching", "A Misdemeanor"],
        "170.15": ["Forgery in the first degree", "C Felony"],
        "170.10": ["Forgery in the second degree", "D Felony"],
        "170.05": ["Forgery in the third degree", "A Misdemeanor"],
        "170.65": ["Forgery of a vehicle identification number", "E Felony"],
        "165.35": ["Fortune telling", "B Misdemeanor"],
        "185.00": ["Fraud in insolvency", "A Misdemeanor"],
        "185.05": ["Fraud involving a security interest", "A Misdemeanor"],
        "165.30": ["Fraudulent accosting", "A Misdemeanor"],
        "185.10": ["Fraudulent disposition of mortgaged property", "A Misdemeanor"],
        "185.15": ["Fraudulent disposition of property subject to a conditional sale contract", "A Misdemeanor"],
        "170.25": ["Fraudulent making of an electronic access device in the second degree", "D Felony"],
        "165.20": ["Fraudulently obtaining a signature", "A Misdemeanor"],
        "225.60": ["Gaming fraud in the first degree", "E Felony"],
        "225.55": ["Gaming fraud in the second degree", "A Misdemeanor"],
        "120.07": ["Gang assault in the first degree", "B Violent Felony"],
        "120.06": ["Gang assault in the second degree", "C Violent Felony"],
        "200.30": ["Giving unlawful gratuities", "A Misdemeanor"],
        "155.42": ["Grand larceny in the first degree", "B Felony"],
        "155.30": ["Grand larceny in the fourth degree", "E Felony"],
        "155.40": ["Grand larceny in the second degree", "C Felony"],
        "155.35": ["Grand larceny in the third degree", "D Felony"],
        "240.25": ["Harassment in the first degree", "B Misdemeanor"],
        "240.26": ["Harassment in the second degree", "Violation"],
        "241.05": ["Harassment of a rent regulated tenant", "E Felony"],
        "242.15": ["Harming a service animal in the first degree", "E Felony"],
        "242.10": ["Harming a service animal in the second degree", "A Misdemeanor"],
        "195.12": ["Harming an animal trained to aid a person with a disability in the first degree", "A Misdemeanor"],
        "195.11": ["Harming an animal trained to aid a person with a disability in the second degree", "B Misdemeanor"],
        "120.16": ["Hazing in the first degree", "A Misdemeanor"],
        "120.17": ["Hazing in the second degree", "Violation"],
        "177.05": ["Health care fraud in the fifth degree", "A Misdemeanor"],
        "177.25": ["Health care fraud in the first degree", "B Felony"],
        "177.10": ["Health care fraud in the fourth degree", "E Felony"],
        "177.20": ["Health care fraud in the second degree", "C Felony"],
        "177.15": ["Health care fraud in the third degree", "D Felony"],
        "205.65": ["Hindering prosecution in the first degree", "D Felony"],
        "205.60": ["Hindering prosecution in the second degree", "E Felony"],
        "205.55": ["Hindering prosecution in the third degree", "A Misdemeanor"],
        "490.35": ["Hindering prosecution of terrorism in the first degree", "B Violent Felony"],
        "490.30": ["Hindering prosecution of terrorism in the second degree", "C Violent Felony"],
        "190.80": ["Identity theft in the first degree", "D Felony"],
        "190.79": ["Identity theft in the second degree", "E Felony"],
        "190.78": ["Identity theft in the third degree", "A Misdemeanor"],
        "170.70": ["Illegal possession of a vehicle identification number", "E Felony"],
        "190.89": ["Immigrant assistance services fraud in the first degree", "E Felony"],
        "190.87": ["Immigrant assistance services fraud in the second degree", "A Misdemeanor"],
        "200.55": ["Impairing the integrity of a government licensing examination", "D Felony"],
        "180.52": ["Impairing the integrity of a pari-mutuel betting system in the second degree", "E Felony"],
        "255.27": ["Incest in the first degree", "B Violent Felony"],
        "255.26": ["Incest in the second degree", "D Felony"],
        "255.25": ["Incest in the third degree", "E Felony"],
        "240.08": ["Inciting to riot", "A Misdemeanor"],
        "176.10": ["Insurance fraud in the fifth degree", "A Misdemeanor"],
        "176.30": ["Insurance fraud in the first degree", "B Felony"],
        "176.15": ["Insurance fraud in the fourth degree", "E Felony"],
        "176.25": ["Insurance fraud in the second degree", "C Felony"],
        "176.20": ["Insurance fraud in the third degree", "D Felony"],
        "242.05": ["Interference harassment or intimidation of a service animal", "B Misdemeanor"],
        "215.17": ["Intimidating a victim or witness in the first degree", "B Violent Felony"],
        "215.16": ["Intimidating a victim or witness in the second degree", "D Violent Felony"],
        "215.15": ["Intimidating a victim or witness in the third degree", "E Felony"],
        "190.05": ["Issuing a bad check", "B Misdemeanor"],
        "175.40": ["Issuing a false certificate", "E Felony"],
        "175.45": ["Issuing a false financial statement", "A Misdemeanor"],
        "125.60": ["Issuing abortional articles", "B Misdemeanor"],
        "165.25": ["Jostling", "A Misdemeanor"],
        "135.25": ["Kidnapping in the first degree", "A-I Felony"],
        "135.20": ["Kidnapping in the second degree", "B Violent Felony"],
        "195.06": ["Killing or injuring a police animal", "A Misdemeanor"],
        "135.35": ["Labor trafficking", "D Felony"],
        "400.00": ["Licenses to carry possess repair and dispose of firearms", "A Misdemeanor"],
        "176.45": ["Life settlement fraud in the fifth degree", "A Misdemeanor"],
        "176.65": ["Life settlement fraud in the first degree", "B Felony"],
        "176.50": ["Life settlement fraud in the fourth degree", "E Felony"],
        "176.60": ["Life settlement fraud in the second degree", "C Felony"],
        "240.35": ["Loitering", "Violation"],
        "240.37": ["Loitering for the purpose of engaging in a prostitution offense", "A Misdemeanor"],
        "240.37": ["Loitering for the purpose of engaging in a prostitution offense", "Violation"],
        "240.36": ["Loitering in the first degree", "B Misdemeanor"],
        "120.70": ["Luring a child", "E Felony"],
        "190.55": ["Making a false statement of credit terms", "A Misdemeanor"],
        "210.45": ["Making a punishable false written statement", "A Misdemeanor"],
        "490.20": ["Making a terroristic threat", "D Violent Felony"],
        "210.40": ["Making an apparently sworn false statement in the first degree", "E Felony"],
        "210.35": ["Making an apparently sworn false statement in the second degree", "A Misdemeanor"],
        "145.60": ["Making graffiti", "A Misdemeanor"],
        "225.90": ["Manipulation of gaming outcomes at an authorized gaming establishment", "A Misdemeanor"],
        "125.20": ["Manslaughter in the first degree", "B Violent Felony"],
        "125.15": ["Manslaughter in the second degree", "C Felony"],
        "275.10": ["Manufacture of unauthorized recordings in the first degree", "E Felony"],
        "275.05": ["Manufacture of unauthorized recordings in the second degree", "A Misdemeanor"],
        "275.20": ["Manufacture or sale of an unauthorized recording of a performance in the first degree", "E Felony"],
        "275.15": ["Manufacture or sale of an unauthorized recording of a performance in the second degree", "A Misdemeanor"],
        "265.10": ["Manufacture transport disposition and defacement of weapons and dangerous instruments and appliances", "A Misdemeanor"],
        "120.18": ["Menacing a police officer or peace officer", "D Violent Felony"],
        "120.13": ["Menacing in the first degree", "E Felony"],
        "120.14": ["Menacing in the second degree", "A Misdemeanor"],
        "120.15": ["Menacing in the third degree", "B Misdemeanor"],
        "165.00": ["Misapplication of property", "A Misdemeanor"],
        "215.30": ["Misconduct by a juror in the first degree", "A Misdemeanor"],
        "215.28": ["Misconduct by a juror in the second degree", "Violation"],
        "190.35": ["Misconduct by corporate official", "B Misdemeanor"],
        "260.31": ["Misrepresentation by a child day care provider", "A Misdemeanor"],
        "470.24": ["Money laundering in support of terrorism in the first degree", "B Felony"],
        "470.21": ["Money laundering in support of terrorism in the fourth degree", "E Felony"],
        "470.23": ["Money laundering in support of terrorism in the second degree", "C Felony"],
        "470.22": ["Money laundering in support of terrorism in the third degree", "D Felony"],
        "470.20": ["Money laundering in the first degree", "B Felony"],
        "470.05": ["Money laundering in the fourth degree", "E Felony"],
        "470.15": ["Money laundering in the second degree", "C Felony"],
        "470.10": ["Money laundering in the third degree", "D Felony"],
        "125.27": ["Murder in the first degree", "A-I Felony"],
        "125.25": ["Murder in the second degree", "A-I Felony"],
        '260.06': ['Non-support of a child in the first degree', 'E Felony'],
        '260.05': ['Non-support of a child in the second degree', 'A Misdemeanor'],
        '235.07': ['Obscenity in the first degree', 'D Felony'],
        '235.06': ['Obscenity in the second degree', 'E Felony'],
        '235.05': ['Obscenity in the third degree', 'A Misdemeanor'],
        '195.16': ['Obstructing emergency medical services', 'A Misdemeanor'],
        '195.08': ['Obstructing governmental administration by means of a self-defense spray device', 'D Felony'],
        '195.05': ['Obstructing governmental administration in the second degree', 'A Misdemeanor'],
        '195.17': ['Obstruction of governmental duties by means of a bomb destructive device explosive or hazardous substance', 'D Felony'],
        '245.05': ['Offensive exhibition', 'Violation'],
        '175.35': ['Offering a false instrument for filing in the first degree', 'E Felony'],
        '175.30': ['Offering a false instrument for filing in the second degree', 'A Misdemeanor'],
        '195.00': ['Official misconduct', 'A Misdemeanor'],
        '220.77': ['Operating as a major trafficker', 'A-I Felony'],
        '230.06': ['Patronizing a prostitute in the first degree', 'D Felony'],
        '230.05': ['Patronizing a prostitute in the second degree', 'E Felony'],
        '230.04': ['Patronizing a prostitute in the third degree', 'A Misdemeanor'],
        '210.15': ['Perjury in the first degree', 'D Felony'],
        '210.10': ['Perjury in the second degree', 'E Felony'],
        '210.05': ['Perjury in the third degree', 'A Misdemeanor'],
        '230.40': ['Permitting prostitution', 'B Misdemeanor'],
        '130.53': ['Persistent sexual abuse', 'E Violent Felony'],
        '155.25': ['Petit larceny', 'A Misdemeanor'],
        '240.63': ['Placing a false bomb or hazardous substance in a sports stadium or arena mass transportation facility or enclosed shopping mall', 'D Violent Felony'],
        '240.62': ['Placing a false bomb or hazardous substance in the first degree', 'D Violent Felony'],
        '240.61': ['Placing a false bomb or hazardous substance in the second degree', 'E Violent Felony'],
        '263.16': ['Possessing a sexual performance by a child', 'E Felony'],
        '263.11': ['Possessing an obscene sexual performance by a child', 'E Felony'],
        '225.30': ['Possession of a gambling device', 'A Misdemeanor'],
        '140.35': ['Possession of burglar`s tools', 'A Misdemeanor'],
        '250.10': ['Possession of eavesdropping devices', 'A Misdemeanor'],
        '225.20': ['Possession of gambling records in the first degree', 'E Felony'],
        '225.15': ['Possession of gambling records in the second degree', 'A Misdemeanor'],
        '145.65': ['Possession of graffiti instruments', 'B Misdemeanor'],
        '225.80': ['Possession of unlawful gaming property in the first degree', 'D Felony'],
        '225.75': ['Possession of unlawful gaming property in the second degree', 'E Felony'],
        '225.70': ['Possession of unlawful gaming property in the third degree', 'A Misdemeanor'],
        '190.45': ['Possession of usurious loan records', 'A Misdemeanor'],
        '130.95': ['Predatory sexual assault', 'A-II Felony'],
        '130.96': ['Predatory sexual assault against a child.', 'A-II Felony'],
        '265.35': ['Prohibited use of weapons', 'A Misdemeanor'],
        '263.15': ['Promoting a sexual performance by a child', 'D Felony'],
        '120.30': ['Promoting a suicide attempt', 'E Felony'],
        '263.10': ['Promoting an obscene sexual performance by a child', 'D Felony'],
        '225.10': ['Promoting gambling in the first degree', 'E Felony'],
        '225.05': ['Promoting gambling in the second degree', 'A Misdemeanor'],
        '205.25': ['Promoting prison contraband in the first degree', 'D Felony'],
        '205.20': ['Promoting prison contraband in the second degree', 'A Misdemeanor'],
        '230.19': ['Promoting prostitution in a school zone', 'E Felony'],
        '230.32': ['Promoting prostitution in the first degree', 'B Felony'],
        '230.20': ['Promoting prostitution in the fourth degree', 'A Misdemeanor'],
        '230.30': ['Promoting prostitution in the second degree', 'C Felony'],
        '230.25': ['Promoting prostitution in the third degree', 'D Felony'],
        '245.02': ['Promoting the exposure of a person', 'Violation'],
        '230.00': ['Prostitution', 'B Misdemeanor'],
        '230.03': ['Prostitution in a school zone', 'A Misdemeanor'],
        '215.22': ['Providing a juror with a gratuity', 'A Misdemeanor'],
        '245.11': ['Public display of offensive sexual material', 'A Misdemeanor'],
        '245.00': ['Public lewdness', 'B Misdemeanor'],
        '245.03': ['Public lewdness in the first degree', 'A Misdemeanor'],
        '130.35': ['Rape in the first degree', 'B Violent Felony'],
        '130.30': ['Rape in the second degree', 'D Violent Felony'],
        '130.25': ['Rape in the third degree', 'E Felony'],
        '200.27': ['Receiving reward for official misconduct in the first degree', 'C Felony'],
        '200.25': ['Receiving reward for official misconduct in the second degree', 'E Felony'],
        '200.35': ['Receiving unlawful gratuities', 'A Misdemeanor'],
        '120.02': ['Reckless assault of a child', 'D Violent Felony'],
        '120.01': ['Reckless assault of a child by a child day care provider', 'E Felony'],
        '120.25': ['Reckless endangerment in the first degree', 'D Felony'],
        '120.20': ['Reckless endangerment in the second degree', 'A Misdemeanor'],
        '145.25': ['Reckless endangerment of property', 'B Misdemeanor'],
        '195.10': ['Refusing to aid a peace or a police officer', 'B Misdemeanor'],
        '180.57': ['Rent gouging in the first degree', 'E Felony'],
        '180.56': ['Rent gouging in the second degree', 'A Misdemeanor'],
        '180.55': ['Rent gouging in the third degree', 'B Misdemeanor'],
        '187.05': ['Residential mortgage fraud in the fifth degree', 'A Misdemeanor'],
        '187.25': ['Residential mortgage fraud in the first degree', 'B Felony'],
        '187.10': ['Residential mortgage fraud in the fourth degree', 'E Felony'],
        '187.20': ['Residential mortgage fraud in the second degree', 'C Felony'],
        '187.15': ['Residential mortgage fraud in the third degree', 'D Felony'],
        '200.22': ['Rewarding official misconduct in the first degree', 'C Felony'],
        '200.20': ['Rewarding official misconduct in the second degree', 'E Felony'],
        '240.06': ['Riot in the first degree', 'E Felony'],
        '240.05': ['Riot in the second degree', 'A Misdemeanor'],
        '160.15': ['Robbery in the first degree', 'B Violent Felony'],
        '160.10': ['Robbery in the second degree', 'C Violent Felony'],
        '160.05': ['Robbery in the third degree', 'D Felony'],
        "265.45": ["Safe storage of rifles, shotguns, and firearms", "A Misdemeanor"],
        "190.65": ["Scheme to defraud in the first degree", "E Felony"],
        "190.60": ["Scheme to defraud in the second degree", "A Misdemeanor"],
        "190.70": ["Scheme to defraud the state by unlawfully selling prescriptions", "A Misdemeanor"],
        "125.55": ["Self-abortion in the first degree", "A Misdemeanor"],
        "125.50": ["Self-abortion in the second degree", "B Misdemeanor"],
        "400.03": ["Sellers of ammunition", "Violation"],
        "130.65": ["Sexual abuse in the first degree", "D Violent Felony"],
        "130.60": ["Sexual abuse in the second degree", "A Misdemeanor"],
        "130.55": ["Sexual abuse in the third degree", "B Misdemeanor"],
        "130.20": ["Sexual misconduct", "A Misdemeanor"],
        "490.15": ["Soliciting or providing support for an act of terrorism in the first degree", "C Violent Felony"],
        "490.10": ["Soliciting or providing support for an act of terrorism in the second degree", "D Violent Felony"],
        "180.45": ["Sports bribe receiving", "E Felony"],
        "120.60": ["Stalking in the first degree", "D Violent Felony"],
        "120.45": ["Stalking in the fourth degree", "B Misdemeanor"],
        "120.55": ["Stalking in the second degree", "E Felony"],
        "120.50": ["Stalking in the third degree", "A Misdemeanor"],
        "120.13": ["Strangulation in the first degree", "C Violent Felony"],
        "121.12": ["Strangulation in the second degree", "D Violent Felony"],
        "135.55": ["Substitution of children", "E Felony"],
        "145.45": ["Tampering with a consumer product in the first degree", "E Felony"],
        "145.40": ["Tampering with a consumer product in the second degree", "A Misdemeanor"],
        "215.25": ["Tampering with a juror in the first degree", "A Misdemeanor"],
        "215.23": ["Tampering with a juror in the second degree", "B Misdemeanor"],
        "180.51": ["Tampering with a sports contest in the first degree", "E Felony"],
        "180.50": ["Tampering with a sports contest in the second degree", "A Misdemeanor"],
        "215.13": ["Tampering with a witness in the first degree", "B Felony"],
        "215.10": ["Tampering with a witness in the fourth degree", "A Misdemeanor"],
        "215.12": ["Tampering with a witness in the second degree", "D Felony"],
        "215.11": ["Tampering with a witness in the third degree", "E Felony"],
        "215.40": ["Tampering with physical evidence", "E Felony"],
        "250.25": ["Tampering with private communications", "B Misdemeanor"],
        "175.25": ["Tampering with public records in the first degree", "D Felony"],
        "175.20": ["Tampering with public records in the second degree", "A Misdemeanor"],
        "165.15": ["Theft of services", "A Misdemeanor"],
        "165.72": ["Trademark counterfeiting in the second degree", "E Felony"],
        "165.71": ["Trademark counterfeiting in the third degree", "A Misdemeanor"],
        "140.05": ["Trespass", "Violation"],
        "275.34": ["Unauthorized operation of a recording device in a motion picture or live theater in the first degree", "E Felony"],
        "275.33": ["Unauthorized operation of a recording device in a motion picture or live theater in the second degree", "A Misdemeanor"],
        "190.72": ["Unauthorized radio transmission", "A Misdemeanor"],
        "165.16": ["Unauthorized sale of certain transportation services", "B Misdemeanor"],
        "156.05": ["Unauthorized use of a computer", "A Misdemeanor"],
        "165.08": ["Unauthorized use of a vehicle in the first degree", "D Felony"],
        "165.06": ["Unauthorized use of a vehicle in the second degree", "E Felony"],
        "165.05": ["Unauthorized use of a vehicle in the third degree", "A Misdemeanor"],
        "265.36": ["Unlawful possession of a large capacity ammunition feeding", "A Misdemeanor"],
        "240.10": ["Unlawful assembly", "B Misdemeanor"],
        "190.50": ["Unlawful collection practices", "B Misdemeanor"],
        "215.75": ["Unlawful disclosure of an indictment", "B Misdemeanor"],
        "245.15": ["Unlawful dissemination or publication of an intimate image", "A Misdemeanor"],
        "220.76": ["Unlawful disposal of methamphetamine laboratory material", "E Felony"],
        "215.80": ["Unlawful disposition of assets subject to forfeiture", "A Misdemeanor"],
        "156.30": ["Unlawful duplication of computer related material in the first degree", "E Felony"],
        "156.29": ["Unlawful duplication of computer related material in the second degree", "B Misdemeanor"],
        "270.35": ["Unlawful fleeing a police officer in a motor vehicle in the first degree", "D Felony"],
        "270.30": ["Unlawful fleeing a police officer in a motor vehicle in the second degree", "E Felony"],
        "270.25": ["Unlawful fleeing a police officer in a motor vehicle in the third degree", "A Misdemeanor"],
        "215.70": ["Unlawful grand jury disclosure", "E Felony"],
        "135.10": ["Unlawful imprisonment in the first degree", "E Felony"],
        "135.05": ["Unlawful imprisonment in the second degree", "A Misdemeanor"],
        "220.75": ["Unlawful manufacture of methamphetamine in the first degree", "B Felony"],
        "220.74": ["Unlawful manufacture of methamphetamine in the second degree", "C Felony"],
        "220.73": ["Unlawful manufacture of methamphetamine in the third degree", "D Felony"],
        "225.95": ["Unlawful manufacture, sale, distribution, marking, altering or modification of equipment and devices associated with gaming", "A Misdemeanor"],
        "190.86": ["Unlawful possession of a skimmer device in the first degree", "E Felony"],
        "190.85": ["Unlawful possession of a skimmer device in the second degree", "A Misdemeanor"],
        "265.37": ["Unlawful possession of certain ammunition feeding devices", "B Misdemeanor"],
        "221.05": ["Unlawful possession of marihuana", "Violation"],
        "190.83": ["Unlawful possession of personal identification information in the first degree", "D Felony"],
        "190.82": ["Unlawful possession of personal identification information in the second degree", "E Felony"],
        "190.81": ["Unlawful possession of personal identification information in the third degree", "A Misdemeanor"],
        "140.40": ["Unlawful possession of radio devices", "B Misdemeanor"],
        "240.65": ["Unlawful prevention of public access to records", "Violation"],
        "222.45": ["Unlawful sale of cannabis", "Violation"],
        "250.50": ["Unlawful surveillance in the first degree", "D Felony"],
        "250.45": ["Unlawful surveillance in the second degree", "E Felony"],
        "165.17": ["Unlawful use of credit card debit card or public benefit card", "A Misdemeanor"],
        "165.07": ["Unlawful use of secret scientific material", "E Felony"],
        "270.20": ["Unlawful wearing of a body vest", "E Felony"],
        "190.30": ["Unlawfully concealing a will", "E Felony"],
        "260.20": ["Unlawfully dealing with a child in the first degree", "A Misdemeanor"],
        "260.21": ["Unlawfully dealing with a child in the second degree", "B Misdemeanor"],
        "270.00": ["Unlawfully dealing with fireworks and dangerous fireworks", "B Misdemeanor"],
        "255.05": ['Unlawfully issuing a dissolution decree', 'A Misdemeanor'],
        "270.05": ['Unlawfully possessing or selling noxious material', 'B Misdemeanor'],
        "145.30": ['Unlawfully posting advertisements', 'Violation'],
        "255.10": ['Unlawfully procuring a marriage license', 'A Misdemeanor'],
        "270.15": ['Unlawfully refusing to yield a party line', 'B Misdemeanor'],
        "255.00": ['Unlawfully solemnizing a marriage', 'A Misdemeanor'],
        "170.60": ['Unlawfully using slugs in the first degree', 'E Felony'],
        "170.55": ['Unlawfully using slugs in the second degree', 'B Misdemeanor'],
        "405.12": ['Unpermitted use of indoor pyrotechnics in the second degree', 'A Misdemeanor'],
        "263.05": ['Use of a child in a sexual performance', 'C Felony'],
        "220.28": ['Use of a child to commit a controlled substance offense', 'E Felony'],
        "225.65": ['Use of counterfeit, unapproved or unlawful wagering instruments', 'A Misdemeanor'],
        "225.85": ['Use of unlawful gaming property', 'E Felony'],
        "120.04": ['Vehicular assault in the first degree', 'D Felony'],
        "120.03": ['Vehicular assault in the second degree', 'E Felony'],
        "125.13": ['Vehicular manslaughter in the first degree', 'C Felony'],
        "125.12": ['Vehicular manslaughter in the second degree', 'D Felony'],
        "158.05": ['Welfare fraud in the fifth degree', 'A Misdemeanor'],
        "158.25": ['Welfare fraud in the first degree', 'B Felony'],
        "158.10": ['Welfare fraud in the fourth degree', 'E Felony'],
        "158.20": ['Welfare fraud in the second degree', 'C Felony'],
        "158.15": ['Welfare fraud in the third degree', 'D Felony'],
        "205.30": ['Offenses against public admini', 'A Misdemeanor'],
        "121.13": ['Felony Assault', 'B Felony'],
        "195.07": ['Miscellaneous penal law', 'E Felony'],
        "230.34": [ 'Compelling prostitution',  'B Felony'],
        "165.73": ['Forgery of a vehicle identification number',  'E Felony']

    }

    def format_law_code(law_code):
        if law_code.startswith('PL'):
            stripped_code = law_code.replace('PL', '').strip()
            first_three = stripped_code[:3]
            next_two = stripped_code[3:5]
            formatted_code = f'{first_three}.{next_two}'
            return formatted_code
        return None


    data['Formatted_Code'] = data['Law_Code'].apply(format_law_code)


    def fetch_description_and_class(formatted_code):
        if formatted_code and formatted_code in offense_data:
            return pd.Series(offense_data[formatted_code], index=['Description', 'Class'])
        else:
            return pd.Series(['Unknown', 'Infraction'], index=['Description', 'Class'])


    data[['Description', 'Class']] = data['Formatted_Code'].apply(fetch_description_and_class)


    data['Description'] = data['Description'].where(data['Description'] != 'Unknown', data['Offense_Description'])


    del data['Formatted_Code']
    return data


  def determine_indicator(self, row):
    class_code = row['Class_Code']
    days = row['No_Days']

    if class_code == 'F':
        if days <= 8:
            return 'Red'
        elif 8 < days <= 30:
            return 'Orange'
        elif 30 < days <= 90:
            return 'Yellow'
        else:
            return 'Blue'
    elif class_code == 'M':
        if days <= 8:
            return 'Orange'
        elif 8 < days <= 30:
            return 'Yellow'
        else:
            return 'Blue'
    elif class_code == 'V':
        if days <= 8:
            return 'Yellow'
        elif 8 < days <= 30:
            return 'Blue'
        else:
            return 'Green'
    elif class_code == 'I':
        if days <= 8:
            return 'Blue'
        else:
            return 'Green'
    return 'White'


  def create_indicator(self, filtered_df):
    def extract_class_code(class_str):
        class_list = class_str.split()
        return class_list[-1][0]


    filtered_df['Class_Code'] = filtered_df['Class'].apply(extract_class_code)
    filtered_df['Class_Code'] = filtered_df['Class_Code'].fillna('I')


    filtered_df['Arrest_Date'] = pd.to_datetime(filtered_df['Arrest_Date'], errors='coerce')


    valid_dates = filtered_df['Arrest_Date'].notna()
    filtered_df = filtered_df[valid_dates]


    most_recent_date = filtered_df['Arrest_Date'].max()
    filtered_df['No_Days'] = (most_recent_date - filtered_df['Arrest_Date']).dt.days


    filtered_df['Indicator'] = filtered_df.apply(self.determine_indicator, axis=1)

    category_map = {
        'F' : 'Felony',
        'M' : 'Misdemeanor',
        'V' : 'Violation',
        'I' : 'Infraction'
    }
    filtered_df['Class_Code'] = filtered_df['Class_Code'].replace(category_map)
    return filtered_df

  def datetime_columns(self, data):
    # Convert the 'date' column to datetime
    data['Arrest_Date'] = pd.to_datetime(data['Arrest_Date'])

    # Extract year, month, day, and day of the week
    data['year'] = data['Arrest_Date'].dt.year
    # data['month'] = data['Arrest_Date'].dt.month
    data['day'] = data['Arrest_Date'].dt.day
    # data['day_of_week'] = data['Arrest_Date'].dt.dayofweek
    data['day_name'] = data['Arrest_Date'].dt.day_name()
    data['month_name'] = data['Arrest_Date'].dt.month_name()
    return data

  def rename_values(self, df):
    df['Perpetrator_Sex'] = df['Perpetrator_Sex'].replace({'M': 'Male', 'F': 'Female'})
    borough_map = {
        'M': 'Manhattan',
        'K': 'Brooklyn (Kings County)',
        'Q': 'Queens',
        'B': 'The Bronx',
        'S': 'Staten Island'
    }

    df['Arrest_Borough'] = df['Arrest_Borough'].replace(borough_map)
    return df




# 1. Loading Data:

In [ ]:
data = pd.read_csv("/content/drive/MyDrive/Capstone Project/NYPD_Arrest_Data__Year_to_Date__20240612.csv")


# 2. Inspecting Data:

In [ ]:
crime = CrimeHotspot()

1. **index**: An auto-incrementing index value for each row in the dataset.
2. **Arrest_ID**: A unique identifier for each arrest.
3. **Arrest_Date**: The date on which the arrest occurred.
4. **PD_Code**: The Police Department code associated with the offense.
5. **PD_Description**: A textual description of the offense corresponding to the PD_Code.
6. **KY_Code**: The New York State Penal Law Code associated with the offense.
7. **Offense_Description**: A detailed description of the offense.
8. **Law_Code**: A code representing the specific law that was violated.
9. **Law_Category**: The category of the law that was violated (e.g., felony, misdemeanor).
10. **Arrest_Borough**: The borough in New York City where the arrest took place.
11. **Arrest_Precinct**: The precinct in New York City where the arrest took place.
12. **Jurisdiction_Code**: A code representing the jurisdiction of the arrest.
13. **Age_Group**: The age group of the individual arrested (e.g., <18, 18-24, 25-44, 45-64, 65+).
14. **Perpetrator_Sex**: The sex of the individual arrested.
15. **Perpetrator_Race**: The race of the individual arrested.
16. **X_Coordinate**: The X coordinate of the location where the arrest occurred (used for mapping).
17. **Y_Coordinate**: The Y coordinate of the location where the arrest occurred (used for mapping).
18. **Latitude**: The latitude of the location where the arrest occurred.
19. **Longitude**: The longitude of the location where the arrest occurred.
20. **Georeferenced_Location**: A georeferenced point combining the latitude and longitude for mapping purposes.

In [ ]:
crime.get_info(data)

Basic information about dataset
-------------------------------------------------------------------------------------------
Number of Rows : 63621

RangeIndex(start=0, stop=63621, step=1)

Number of columns :20

column Name : ['index' 'ARREST_KEY' 'ARREST_DATE' 'PD_CD' 'PD_DESC' 'KY_CD' 'OFNS_DESC'
 'LAW_CODE' 'LAW_CAT_CD' 'ARREST_BORO' 'ARREST_PRECINCT'
 'JURISDICTION_CODE' 'AGE_GROUP' 'PERP_SEX' 'PERP_RACE' 'X_COORD_CD'
 'Y_COORD_CD' 'Latitude' 'Longitude' 'New Georeferenced Column']

Columns which has null values :  {'PD_CD': 2, 'KY_CD': 9, 'LAW_CAT_CD': 402}

-------------------------------------------------------------------------------------------


- The dataset consists of 63,621 rows and 20 columns.
- The columns included are: 'index', 'ARREST_KEY', 'ARREST_DATE', 'PD_CD', 'PD_DESC', 'KY_CD', 'OFNS_DESC', 'LAW_CODE', 'LAW_CAT_CD', 'ARREST_BORO', 'ARREST_PRECINCT', 'JURISDICTION_CODE', 'AGE_GROUP', 'PERP_SEX', 'PERP_RACE', 'X_COORD_CD', 'Y_COORD_CD', 'Latitude', 'Longitude', and 'New Georeferenced Column'.
- There are 3 columns with null values:
  * 'PD_CD' contains 2 null values.
  * KY_CD' contains 9 null values.
  * 'LAW_CAT_CD' contains 402 null values.
- The index range starts from 0 and goes up to 63,621 with a step of 1.

In [ ]:

data.head(5)

ARREST_KEY ARREST_DATE  PD_CD                         PD_DESC  KY_CD  \
0   280255493  01/10/2024  397.0  ROBBERY,OPEN AREA UNCLASSIFIED  105.0   
1   279805419  01/02/2024  157.0                          RAPE 1  104.0   
2   279895750  01/03/2024  101.0                       ASSAULT 3  344.0   
3   280809090  01/19/2024  511.0  CONTROLLED SUBSTANCE, POSSESSI  235.0   
4   280357135  01/11/2024  109.0        ASSAULT 2,1,UNCLASSIFIED  106.0   

                      OFNS_DESC    LAW_CODE LAW_CAT_CD ARREST_BORO  \
0                       ROBBERY  PL 1601001          F           M   
1                          RAPE  PL 1303501          F           K   
2  ASSAULT 3 & RELATED OFFENSES  PL 1200001          M           Q   
3               DANGEROUS DRUGS  PL 2200300          M           B   
4                FELONY ASSAULT  PL 1200502          F           K   

   ARREST_PRECINCT  JURISDICTION_CODE AGE_GROUP PERP_SEX       PERP_RACE  \
0               26                  0       <18        M           BLACK   
1               77                  0     25-44        M  WHITE HISPANIC   
2              106                  0       65+        F           WHITE   
3               49                  0     45-64        M           BLACK   
4               81                  2     25-44        F           BLACK   

   X_COORD_CD  Y_COORD_CD   Latitude  Longitude  \
0      996342      236149  40.814845 -73.956312   
1     1003509      185018  40.674496 -73.930571   
2     1026836      180689  40.662526 -73.846499   
3     1027430      251104  40.855793 -73.843908   
4     1002457      192292  40.694456 -73.934343   

                     New Georeferenced Column  
0                POINT (-73.956312 40.814845)  
1  POINT (-73.9305713255961 40.6744956865259)  
2                POINT (-73.846499 40.662526)  
3                POINT (-73.843908 40.855793)  
4                POINT (-73.934343 40.694456)

In [ ]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 63621 entries, 0 to 63620
Data columns (total 19 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   ARREST_KEY                63621 non-null  int64  
 1   ARREST_DATE               63621 non-null  object 
 2   PD_CD                     63619 non-null  float64
 3   PD_DESC                   63621 non-null  object 
 4   KY_CD                     63612 non-null  float64
 5   OFNS_DESC                 63621 non-null  object 
 6   LAW_CODE                  63621 non-null  object 
 7   LAW_CAT_CD                63219 non-null  object 
 8   ARREST_BORO               63621 non-null  object 
 9   ARREST_PRECINCT           63621 non-null  int64  
 10  JURISDICTION_CODE         63621 non-null  int64  
 11  AGE_GROUP                 63621 non-null  object 
 12  PERP_SEX                  63621 non-null  object 
 13  PERP_RACE                 63621 non-null  object 
 14  X_COOR

In [ ]:
data.describe()

ARREST_KEY         PD_CD         KY_CD  ARREST_PRECINCT  \
count  6.362100e+04  63619.000000  63612.000000     63621.000000   
mean   2.822203e+08    440.414687    254.941002        63.337279   
std    1.387821e+06    274.537356    150.296448        35.192684   
min    2.797635e+08      2.000000    101.000000         1.000000   
25%    2.810329e+08    139.000000    113.000000        40.000000   
50%    2.822323e+08    419.000000    341.000000        63.000000   
75%    2.834321e+08    705.000000    344.000000       101.000000   
max    2.846252e+08    973.000000    995.000000       123.000000   

       JURISDICTION_CODE    X_COORD_CD     Y_COORD_CD      Latitude  \
count       63621.000000  6.362100e+04   63621.000000  63621.000000   
mean            0.946731  1.005320e+06  207682.477515     40.735404   
std             7.187989  2.219886e+04   29489.571873      0.242299   
min             0.000000  0.000000e+00       0.000000      0.000000   
25%             0.000000  9.905030e+05  185710.000000     40.676347   
50%             0.000000  1.005040e+06  206851.000000     40.734434   
75%             0.000000  1.017689e+06  235265.000000     40.812415   
max            97.000000  1.067185e+06  271282.000000     40.911236   

          Longitude  
count  63621.000000  
mean     -73.921505  
std        0.421644  
min      -74.252711  
25%      -73.977428  
50%      -73.924895  
75%      -73.879171  
max        0.000000

# Renaming and Deleting Unwanted Columns:

In [ ]:
data = crime.rename_col(data)
data = crime.rename_values(data)
data = crime.delete_col(data)
column_names = data.columns.tolist()
print(column_names)

['Arrest_ID', 'Arrest_Date', 'Offense_Description', 'Law_Code', 'Law_Category', 'Arrest_Borough', 'Arrest_Precinct', 'Age_Group', 'Perpetrator_Sex', 'Perpetrator_Race', 'X_Coordinate', 'Y_Coordinate', 'Latitude', 'Longitude', 'Georeferenced_Location']


- The dataset has been refined by renaming columns for better understanding and clarity.
- Unnecessary columns such as 'PD_CD', 'PD_DESC', 'KY_CD', and 'JURISDICTION_CODE' have been deleted to streamline the dataset.

# Creating Features:

In [ ]:
data = crime.datetime_columns(data)

# Display the DataFrame with the new columns
column_names = data.columns.tolist()
print(column_names)

['Arrest_ID', 'Arrest_Date', 'Offense_Description', 'Law_Code', 'Law_Category', 'Arrest_Borough', 'Arrest_Precinct', 'Age_Group', 'Perpetrator_Sex', 'Perpetrator_Race', 'X_Coordinate', 'Y_Coordinate', 'Latitude', 'Longitude', 'Georeferenced_Location', 'year', 'day', 'day_name', 'month_name']


- Using the 'Arrest_Date' column, new columns 'year', 'day', 'day_name', and 'month_name' have been created to enhance temporal analysis and visualization of the dataset.
- For example:
  - **Year**: Provides the year of each arrest, facilitating yearly trends analysis.
  - **Day**: Indicates the day of the month when the arrest occurred, aiding in understanding monthly patterns.
  - **Day Name**: Specifies the name of the day (e.g., Monday, Tuesday) for day-specific analysis.
  - **Month Name**: Shows the name of the month (e.g., January, February), useful for monthly trends and seasonality analysis.

In [ ]:
data = crime.create_col(data)
# Display the DataFrame with the new columns
column_names = data.columns.tolist()
print(column_names)

['Arrest_ID', 'Arrest_Date', 'Offense_Description', 'Law_Code', 'Law_Category', 'Arrest_Borough', 'Arrest_Precinct', 'Age_Group', 'Perpetrator_Sex', 'Perpetrator_Race', 'X_Coordinate', 'Y_Coordinate', 'Latitude', 'Longitude', 'Georeferenced_Location', 'year', 'day', 'day_name', 'month_name', 'Description', 'Class']


- Two new columns, 'Description' and 'Class', have been added to the dataset based on the official "ALL Criminal Offenses in NY Penal Code" documentation and 'Law_Code'.
- The 'Description' column standardizes and clarifies offense descriptions, replacing varied and potentially unclear descriptions with official definitions.
- The 'Class' column categorizes offenses according to the NY Penal Code, ensuring accurate classification and correcting any misclassifications.
- These additions aim to enhance dataset clarity, improve model performance by reducing ambiguity, and provide reliable insights into criminal offense analysis.

Reference: https://ypdcrime.com/penallawlist.php



In [ ]:
data = crime.create_indicator(data)
# Display the DataFrame with the new columns
column_names = data.columns.tolist()
print(column_names)

['Arrest_ID', 'Arrest_Date', 'Offense_Description', 'Law_Code', 'Law_Category', 'Arrest_Borough', 'Arrest_Precinct', 'Age_Group', 'Perpetrator_Sex', 'Perpetrator_Race', 'X_Coordinate', 'Y_Coordinate', 'Latitude', 'Longitude', 'Georeferenced_Location', 'year', 'day', 'day_name', 'month_name', 'Description', 'Class', 'Class_Code', 'No_Days', 'Indicator']


1. A new column "days" is added to the dataset, showing how many days ago each crime happened, counting from the most recent date in the dataset.
2. A new column called "Indicator" is added to the dataset to detect crime hotspots. The values are assigned as follows:
  - For Felony crimes:
    * If the crime happened less than 8 days ago, the indicator is RED.
    * If the crime happened between 8 days and 1 month ago, the indicator is ORANGE.
    * If the crime happened between 1 and 3 months ago, the indicator is YELLOW.
    * If the crime happened more than 3 months ago, the indicator is LIGHT BLUE (indicating a serious crime that should disappear after 6 months).
  - For Misdemeanor crimes:
    * If the crime happened less than 8 days ago, the indicator is ORANGE.
    * If the crime happened between 8 days and 1 month ago, the indicator is YELLOW.
    * If the crime happened between 1 and 3 months ago, the indicator is LIGHT BLUE (indicating a crime that should disappear after 3 months).
  - For Violation crimes:
    * If the crime happened less than 8 days ago, the indicator is YELLOW.
    * If the crime happened between 8 days and 1 month ago, the indicator is LIGHT BLUE (indicating a violation that should disappear after 1 month).
  - For Infraction crimes:
    * If the crime happened less than 8 days ago, the indicator is LIGHT BLUE (indicating an infraction that should disappear after 1 month).

In [ ]:
del data['Offense_Description']
del data['Law_Category']

# Handling Duplicates:




In [ ]:
# Count total duplicate rows
total_duplicates = data.duplicated().sum()

print(f"Total Duplicate Rows: {total_duplicates}")

Total Duplicate Rows: 0


# Handling Outliers:

In [ ]:
columns_to_check = [  'Arrest_Borough',
                    'Arrest_Precinct', 'Age_Group', 'Perpetrator_Sex', 'Perpetrator_Race',
                     'year', 'day', 'day_name', 'month_name',
                   'Class_Code', 'Indicator']

# Get unique values for each column
unique_values = {}
for column in columns_to_check:
    unique_values[column] = pd.unique(data[column])

# Print unique values for each column
for column, values in unique_values.items():
    print(f"Unique values for '{column}':")
    print(values)
    print()


Unique values for 'Arrest_Borough':
['Manhattan' 'Brooklyn (Kings County)' 'Queens' 'The Bronx'
 'Staten Island']

Unique values for 'Arrest_Precinct':
[ 26  77 106  49  81 113   6   5  41 102  69  79  20  47 114  88  17 112
  14  45  32 115   1  67 120 103  23  63  40 108  18  46  44  60  52 110
 105  48 121  83 101  43  68 104  42  72  25  71  62  73  75  84  24  94
  66  78  33  34   7  13  30 109  10 107  19  70  90  50 111  28 122   9
  76  61 100 123  22]

Unique values for 'Age_Group':
['<18' '25-44' '65+' '45-64' '18-24']

Unique values for 'Perpetrator_Sex':
['Male' 'Female']

Unique values for 'Perpetrator_Race':
['BLACK' 'WHITE HISPANIC' 'WHITE' 'ASIAN / PACIFIC ISLANDER' 'UNKNOWN'
 'BLACK HISPANIC' 'AMERICAN INDIAN/ALASKAN NATIVE']

Unique values for 'year':
[2024]

Unique values for 'day':
[10  2  3 19 11  1 30 23 31 26  7 15 20 29 16  4 18  6  8 13 22 28 27 12
 25  9 24 17  5 14 21]

Unique values for 'day_name':
['Wednesday' 'Tuesday' 'Friday' 'Thursday' 'Saturday' 'Mond

# Saving the Cleaned Data

In [ ]:
data.to_csv("/content/drive/MyDrive/Capstone Project/Crime_data.csv", index=False)

# Sumarizing Preprocessing steps:

In [ ]:
crime_data = pd.read_csv("/content/drive/MyDrive/Capstone Project/Crime_data.csv")

In [ ]:
crime.get_info(crime_data)

Basic information about dataset
-------------------------------------------------------------------------------------------
Number of Rows : 63621

RangeIndex(start=0, stop=63621, step=1)

Number of columns :23

column Name : ['index' 'Arrest_ID' 'Arrest_Date' 'Law_Code' 'Arrest_Borough'
 'Arrest_Precinct' 'Age_Group' 'Perpetrator_Sex' 'Perpetrator_Race'
 'X_Coordinate' 'Y_Coordinate' 'Latitude' 'Longitude'
 'Georeferenced_Location' 'year' 'day' 'day_name' 'month_name'
 'Description' 'Class' 'Class_Code' 'No_Days' 'Indicator']

Columns which has null values :  {}

-------------------------------------------------------------------------------------------


In [ ]:
crime_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 63621 entries, 0 to 63620
Data columns (total 22 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   Arrest_ID               63621 non-null  int64  
 1   Arrest_Date             63621 non-null  object 
 2   Law_Code                63621 non-null  object 
 3   Arrest_Borough          63621 non-null  object 
 4   Arrest_Precinct         63621 non-null  int64  
 5   Age_Group               63621 non-null  object 
 6   Perpetrator_Sex         63621 non-null  object 
 7   Perpetrator_Race        63621 non-null  object 
 8   X_Coordinate            63621 non-null  int64  
 9   Y_Coordinate            63621 non-null  int64  
 10  Latitude                63621 non-null  float64
 11  Longitude               63621 non-null  float64
 12  Georeferenced_Location  63621 non-null  object 
 13  year                    63621 non-null  int64  
 14  day                     63621 non-null

In [ ]:
crime_data.head(5)

Arrest_ID Arrest_Date    Law_Code           Arrest_Borough  \
0  280255493  2024-01-10  PL 1601001                Manhattan   
1  279805419  2024-01-02  PL 1303501  Brooklyn (Kings County)   
2  279895750  2024-01-03  PL 1200001                   Queens   
3  280809090  2024-01-19  PL 2200300                The Bronx   
4  280357135  2024-01-11  PL 1200502  Brooklyn (Kings County)   

   Arrest_Precinct Age_Group Perpetrator_Sex Perpetrator_Race  X_Coordinate  \
0               26       <18            Male            BLACK        996342   
1               77     25-44            Male   WHITE HISPANIC       1003509   
2              106       65+          Female            WHITE       1026836   
3               49     45-64            Male            BLACK       1027430   
4               81     25-44          Female            BLACK       1002457   

   Y_Coordinate  ...                      Georeferenced_Location  year day  \
0        236149  ...                POINT (-73.956312 40.814845)  2024  10   
1        185018  ...  POINT (-73.9305713255961 40.6744956865259)  2024   2   
2        180689  ...                POINT (-73.846499 40.662526)  2024   3   
3        251104  ...                POINT (-73.843908 40.855793)  2024  19   
4        192292  ...                POINT (-73.934343 40.694456)  2024  11   

    day_name  month_name                                        Description  \
0  Wednesday     January                       Robbery in the second degree   
1    Tuesday     January                           Rape in the first degree   
2  Wednesday     January                        Assault in the third degree   
3     Friday     January  Criminal possession of a controlled substance ...   
4   Thursday     January                       Assault in the second degree   

              Class   Class_Code No_Days Indicator  
0  C Violent Felony       Felony      81    Yellow  
1  B Violent Felony       Felony      89    Yellow  
2     A Misdemeanor  Misdemeanor      88      Blue  
3     A Misdemeanor  Misdemeanor      72      Blue  
4  D Violent Felony       Felony      80    Yellow  

[5 rows x 22 columns]